cloudinit/handlers/shell_script_by_frequency.py

```diff
@@ -0,0 +1,88 @@
+# higher level or in a more general module -- eg maybe in cloudinit/settings.py
+# itself -- but for now it's here.
+pathMap = {
+    PER_ALWAYS: 'per-boot',
+    PER_INSTANCE: 'per-instance',
+    PER_ONCE: 'per-once'
+}
+
+
+# Using pathMap (defined above), return the frequency-specific subfolder for a
+# given frequency constant and parent folder.
+def get_script_folder_by_frequency(freq, scripts_dir):
+    freqPath = pathMap[freq]
+    folder = os.path.join(scripts_dir, freqPath)
+    return folder
+
+
+# Given a filename, a payload, a frequency, and a scripts folder, write the
+# payload to the correct frequency-specific paths
+def write_script_by_frequency(script_path, payload, frequency, scripts_dir):
+    filename = os.path.basename(script_path)
+    filename = util.clean_filename(filename)
+    folder = get_script_folder_by_frequency(frequency, scripts_dir)
+    path = os.path.join(folder, filename)
+    payload = util.dos2unix(payload)
+    util.write_file(path, payload, 0o700)
+
+
+# This is purely to allow packaging args up into a single object to please
+# pylint and avoid 'too many positional args' complaints. I'd be happy to
+# have an alernative.
+class ShellScriptByFreqPartHandler(Handler):
+    def __init__(self, paths, freq, **_kwargs):
+        Handler.__init__(self, freq)
+        self.scripts_dir = paths.get_cpath('scripts')
+        if 'script_path' in _kwargs:
+            self.scripts_dir = paths.get_cpath(_kwargs['script_path'])
+
+    def handle_part(self, data, ctype, script_path, payload, frequency):
+        if script_path is not None:
+            LOG.debug("script_path=%s", script_path)
+            filename = os.path.basename(script_path)
+            filename = util.clean_filename(filename)
+            write_script_by_frequency(script_path, payload, PER_ALWAYS,
+                                      self.scripts_dir)
+
+
+# per-boot
+class ShellScriptPerBootPartHandler(ShellScriptByFreqPartHandler):
+    prefixes = ["text/x-shellscript-per-boot"]
+
+    def __init__(self, paths, **_kwargs):
+        # pylint: disable=too-many-function-args
+        ShellScriptByFreqPartHandler.__init__(self, paths, PER_ALWAYS,
+                                              **_kwargs)
+
+
+# per-instance
+class ShellScriptPerInstancePartHandler(ShellScriptByFreqPartHandler):
+    prefixes = ["text/x-shellscript-per-instance"]
+
+    def __init__(self, paths, **_kwargs):
+        # pylint: disable=too-many-function-args
+        ShellScriptByFreqPartHandler.__init__(self, paths, PER_INSTANCE,
+                                              **_kwargs)
+
+
+# per-once
+class ShellScriptPerOncePartHandler(ShellScriptByFreqPartHandler):
+    prefixes = ["text/x-shellscript-per-once"]
+
+    def __init__(self, paths, **_kwargs):
+        # pylint: disable=too-many-function-args
+        ShellScriptByFreqPartHandler.__init__(self, paths, PER_ONCE,
+                                              **_kwargs)
```

Comment:
```suggestion
    frequency = PER_ONCE
```

Reply:
I can make these changes, but they seem odd. `frequency` is assigned by the `Handler` c'tor, which all these classes inherit from. Is it necessary in Python for subclasses to assign instance variables, even if they're assigned in a parent class? I wouldn't think so but I have been surprised by Python before.

Reply:
Hmmm, there was one other change that I thought I requested, but I'm not seeing now. These other changes won't make sense without it. It was to change `ShellScriptByFreqPartHandler` to look like this:
```python
class ShellScriptByFreqPartHandler(Handler):
    def __init__(self, paths, **_kwargs):
        Handler.__init__(self, self.frequency)
```

Looking at the `Handler` base class now, I can see how this would be a little confusing given that there is also a `frequency` instance variable there.

Since I'm proposing we define class variables named `frequency` in the subclasses, the call to that `Handler.__init__(self, self.frequency)` is passing a reference to the class variable named frequency which then sets the instance variable named `frequency`.

If that's confusing, I'd be ok with changing the names of the class variables (along with the corresponding `Handler.__init__` call) to something like `freq` so it's not so easy to confuse class variable names with instance variable names.

Does that make sense?

Reply:
Ahhhh I suspected my mediocre Python skills played a role here. You were requesting class variables in the subclasses, not instance variables, although there happen to be instance variables of the same name.

The whole `frequency` handling is confusing to me, since `Handler` takes a single frequency in its c'tor and then maintains it as in ivar, but then `handle_part()` takes a frequency as an argument as well.

Meanwhile had this idea: I just checked in `stages` where these handlers actually get created, and it looks like this:

```
        shellscript_per_boot_handler = ShellScriptPerBootPartHandler(**opts)
        shellscript_per_instance_handler = \
            ShellScriptPerInstancePartHandler(**opts)
        shellscript_per_once_handler = ShellScriptPerOncePartHandler(**opts)
```

In theory, those could all be c'tor invocations of a single class, parameterized on `frequency`, and then the subclasses can go away entirely. There'd still be the apparent redundancy of frequency as an arg to `handle_parts()` but that's ok.

Reply:
Ah, right. Your proposed solution makes sense and sounds like a good idea. Go for it!

Reply:
Thanks! Just committed. I had to add a prefixes collection to the base class, of all 3 prefixes, but that's the only part I was a bit unsure about.